ambry-quota/src/main/java/com/github/ambry/quota/AmbryStorageQuotaEnforcer.java

```diff
@@ -0,0 +1,118 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * A {@link StorageQuotaEnforcer} implementation. It keeps an in-memory map for storage quota and listens on the quota change
+ * to update this in-memory map. It also keeps another in-memory map for storage usage of each container and listens on usage
+ * change from {@link StorageUsageRefresher}.
+ *
+ * This implementation checks whether to throttle the operation only if the operation is {@link QuotaOperation#Post}. And when the
+ * targeted account and container doesn't have a quota specified, it doesn't throttle the operation at all. Any legitimate
+ * uploads would also increase the storage usage in the in-memory map.
+ */
+public class AmbryStorageQuotaEnforcer implements StorageQuotaEnforcer {
+  private volatile QuotaMode mode = QuotaMode.Tracking;
+  private volatile Map<String, Map<String, Long>> storageQuota;
+  private volatile Map<String, Map<String, Long>> storageUsage;
+
+  @Override
+  public boolean shouldThrottle(short accountId, short containerId, QuotaOperation op, long size) {
+    if (op != QuotaOperation.Post) {
+      return false;
+    }
+    long quota = storageQuota.getOrDefault(String.valueOf(accountId), new HashMap<>())
+        .getOrDefault(String.valueOf(containerId), Long.MAX_VALUE);
+
+    AtomicBoolean exceedQuota = new AtomicBoolean(false);
+    storageUsage.computeIfAbsent(String.valueOf(accountId), k -> new ConcurrentHashMap<>())
+        .compute(String.valueOf(containerId), (k, v) -> {
+          if (v == null) {
+            return size;
+          }
+          if (v + size < quota) {
+            return v + size;
+          } else {
+            exceedQuota.set(true);
+            return v;
+          }
+        });
+    return mode == QuotaMode.Throttling ? exceedQuota.get() : false;
+  }
+
+  @Override
+  public void setQuotaMode(QuotaMode mode) {
+    this.mode = mode;
+  }
+
+  @Override
+  public void initStorageUsage(Map<String, Map<String, Long>> usage) {
+    storageUsage = new ConcurrentHashMap<>();
+    initMap(usage, storageUsage, true);
+  }
+
+  @Override
+  public void initStorageQuota(Map<String, Map<String, Long>> quota) {
+    storageQuota = new HashMap<>();
+    initMap(quota, storageQuota, false);
+  }
+
+  @Override
+  public StorageUsageRefresher.Listener getUsageRefresherListener() {
+    return new StorageUsageRefresher.Listener() {
+      @Override
+      public void onNewContainerStorageUsage(Map<String, Map<String, Long>> containerStorageUsage) {
+        initMap(containerStorageUsage, storageUsage, true);
+      }
+    };
+  }
+
+  @Override
+  public StorageQuotaSource.Listener getQuotaSourceListener() {
```

Comment:
It should be this class' responsibility to register listener to `AmbryStorageQuotaService`. Instead of returning `Listener` objects, this class should create the object and `AmbryStorageQuotaService.registerListener(listener)`. This will make things simple, as the only responsibility of `AmbryStorageQuotaService` will be to called listener that are registered by consumers. Also the `AmbryStorageQuotaEnforcer` doesn't need to do anything if it doesn't want to listen.

Reply:
updated.